expirationd.lua

```diff
@@ -327,48 +390,202 @@ end
+--     index. If there's no index with this name, will throw an error.
+--     Supported index types are TREE and HASH, using other types will result
+--     in an error.
+-- @func[opt] options.iterate_with
+--     Function which returns an iterator object which provides tuples to
+--     check, considering the `start_key`, `process_while` and other options.
+--     There's a default function which can be overriden with this parameter.
+--
+-- Example of function:
+--
+-- ```
+-- local function iterate_with()
+--     return true
+-- end
+-- ```
+-- @number[opt] options.iteration_delay
+--     Max sleep time between iterations (in seconds). It is allowed to pass
+--     an FFI number: `1LL`, `1UL` etc. Default value is 1 sec.
+--     Fiber sleeps min(`tuples_per_iteration` × `full_scan_time` / `space_length`, `iteration_delay`).
+-- @string[opt] options.iterator_type
+--     Type of the iterator to use, as string or box.index constant, for
+--     example, `EQ` or `box.index.EQ`, default is `box.index.ALL`. See more
+--     about index iterators in [index_object:pairs()][1].
+--
+--     [1]: https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_index/pairs/.
+--
+-- @func[opt] options.on_full_scan_complete
+--     Function to call after completing a full scan iteration. Default value
+--     is a function that do nothing.
+-- @func[opt] options.on_full_scan_error
+--     Function to call after terminating a full scan due to an error. Default
+--     value is a function that do nothing.
+--
+-- Example of function:
+--
+-- ```
+-- local function on_full_scan_complete()
+--     pcall(fiber.sleep, 1)
+-- end
+-- ```
+-- @func[opt] options.on_full_scan_start
+--     Function to call before starting a full scan iteration. Default value
+--     is a function that do nothing.
+-- @func[opt] options.on_full_scan_success
+--     Function to call after successfully completing a full scan iteration.
+--     Default value is a function that do nothing.
+-- @func[opt] options.process_expired_tuple
+--     Applied to expired tuples, receives `space_id`, `args`, `tuple` as
+--     arguments. When `process_expired_tuple` is not passed (or `nil` passed),
+--     tuples are removed.
+--
+-- Example of function:
+--
+-- ```
+-- local function put_tuple_to_archive(space_id, args, tuple)
+--     box.space[space_id]:delete{tuple[1]}
+--     local email = get_field(tuple, 2)
+--     if args.archive_space_id ~= nil and email ~= nil then
+--         box.space[args.archive_space_id]:replace{email, fiber.time()}
+--     end
+-- end
+-- ```
+--
+-- @func[opt] options.process_while
+--     Function to call before checking each tuple. If it returns false, the
+--     task will stop until next full scan. Default is a function that always
+--     return `true`.
+--
+-- Example of function:
+--
+-- ```
+-- local function process_while()
+--     return false
+-- end
+-- ```
+--
+-- @string[opt] options.start_key
```

Comment:
A key may be of any time. I would also say that even if expirationd allows to pass a constant key there, there is no much sense to do so (just bad api design). Any real usage will pass a function there.

Reply:
> A key may be of any time.

Type?


True. Below is a piece of code to process `start_key` argument:
```lua
    -- check start_key                                                           
    if options.start_key ~= nil or options.start_key == box.NULL then            
        if type(options.start_key) == "function" then                            
            task.start_key = function() return options.start_key() end           
        else                                                                     
            task.start_key = function() return options.start_key end             
        end                                                                      
    end      
```

Fixed a type of `start_key` in LDoc comment.

P.S. type of `start_key` in checks() is correct:
```lua
local function expirationd_run_task(name, space_id, is_tuple_expired, options)   
    checks('string', 'number|string', 'function', {                              
        args = '?',                                                              
        ...                                         
        start_key = '?',                                                         
        tuples_per_iteration = '?number|cdata',                                  
        vinyl_assumed_space_len_factor = '?number|cdata',                        
        vinyl_assumed_space_len = '?number|cdata',                               
    })                 
```